### PR TITLE
Relay header and check proof logic

### PR DESCRIPTION
**Pull request by HackFisher**

### CODE REVIEW

File: core/merkle-patricia-trie/src/db.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
+
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/db.rs]: `cargo fmt` please

File: core/merkle-patricia-trie/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/error.rs]: `cargo fmt` please

File: core/merkle-patricia-trie/src/db.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    pub fn contains(&self, key: &[u8]) -> bool {
+        self.data.borrow().contains_key(key)
+    }
+
+    pub fn remove(&self, key: &[u8]) -> Option<Vec<u8>> {
+        self.data.borrow_mut().remove(key)
+    }
+
+    /// Insert a batch of data into the cache.
+    pub fn insert_batch(& self, keys: Vec<Vec<u8>>, values: Vec<Vec<u8>>) {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/db.rs]: `cargo fmt` please

File: core/merkle-patricia-trie/src/db.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        for i in 0..keys.len() {
+            let key = keys[i].clone();
+            let value = values[i].clone();
+            self.insert(key, value);
+        }
+    }
+
+    /// Remove a batch of data into the cache.
+    pub fn remove_batch(& self, keys: &[Vec<u8>]) {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/db.rs]: `cargo fmt` please

File: core/merkle-patricia-trie/src/db.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
+
+use hashbrown::HashMap;
+use crate::std::*;
+
+#[derive(Debug)]
+pub struct MemoryDB {
+    data: RefCell<HashMap<Vec<u8>, Vec<u8>>>
+}
+
+impl MemoryDB {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/db.rs]: recommend add `#[inline]` for these tiny & light functions

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 6]
```diff
@@ -0,0 +1,180 @@
+
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: cargo fmt please

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+use crate::std::*;
+use core::cmp::min;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Nibbles {
+    hex_data: Vec<u8>,
+}
+
+impl Nibbles {
+    pub fn from_hex(hex: Vec<u8>) -> Self {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            hex.push(item % 16);
+        }
+        if is_leaf {
+            hex.push(16);
+        }
+
+        Nibbles { hex_data: hex }
+    }
+
+    pub fn is_leaf(&self) -> bool {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        (raw, is_leaf)
+    }
+
+    pub fn len(&self) -> usize {
+        self.hex_data.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        self.len() == 0
+    }
+
+    pub fn at(&self, i: usize) -> usize {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+        while i < s {
+            if self.at(i) != other_partial.at(i) {
+                break;
+            }
+            i += 1;
+        }
+        i
+    }
+
+    pub fn offset(&self, index: usize) -> Nibbles {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        self.slice(index, self.hex_data.len())
+    }
+
+    pub fn slice(&self, start: usize, end: usize) -> Nibbles {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        Nibbles::from_hex(self.hex_data[start..end].to_vec())
+    }
+
+    pub fn get_data(&self) -> &[u8] {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+    pub fn truncate(&mut self, len: usize) {
+        self.hex_data.truncate(len)
+    }
+
+    pub fn pop(&mut self) -> Option<u8> {
+        self.hex_data.pop()
+    }
+
+    pub fn push(&mut self, e: u8) {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+    pub fn extend(&mut self, b: &Nibbles) {
+        self.hex_data.extend_from_slice(b.get_data());
+    }
+
+    pub fn truncate(&mut self, len: usize) {
+        self.hex_data.truncate(len)
+    }
+
+    pub fn pop(&mut self) -> Option<u8> {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        hex_data.extend_from_slice(self.get_data());
+        hex_data.extend_from_slice(b.get_data());
+        Nibbles::from_hex(hex_data)
+    }
+
+    pub fn extend(&mut self, b: &Nibbles) {
+        self.hex_data.extend_from_slice(b.get_data());
+    }
+
+    pub fn truncate(&mut self, len: usize) {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    pub fn join(&self, b: &Nibbles) -> Nibbles {
+        let mut hex_data = vec![];
+        hex_data.extend_from_slice(self.get_data());
+        hex_data.extend_from_slice(b.get_data());
+        Nibbles::from_hex(hex_data)
+    }
+
+    pub fn extend(&mut self, b: &Nibbles) {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/nibbles.rs [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+
+    pub fn slice(&self, start: usize, end: usize) -> Nibbles {
+        Nibbles::from_hex(self.hex_data[start..end].to_vec())
+    }
+
+    pub fn get_data(&self) -> &[u8] {
+        &self.hex_data
+    }
+
+    pub fn join(&self, b: &Nibbles) -> Nibbles {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/nibbles.rs]: recommend add `#[inline]` for this tiny & light function

File: core/merkle-patricia-trie/src/node.rs [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+#[derive(Debug, Clone)]
+pub enum Node {
+    Empty,
+    Leaf(Rc<RefCell<LeafNode>>),
+    Extension(Rc<RefCell<ExtensionNode>>),
+    Branch(Rc<RefCell<BranchNode>>),
+    Hash(Rc<RefCell<HashNode>>),
+}
+
+impl Node {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/node.rs]: recommend add `#[inline]` for these tiny & light functions

File: core/merkle-patricia-trie/src/proof.rs [hunk 20]
```diff
@@ -0,0 +1,58 @@
+use crate::std::*;
+use rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream};
+
+#[derive(Clone)]
+#[cfg_attr(feature = "std", derive(Debug, PartialEq))]
+pub struct Proof {
+    pub nodes: Vec<Vec<u8>>
+}
+
+impl Proof {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/proof.rs]: recommend add `#[inline]` for these tiny & light functions

File: core/merkle-patricia-trie/src/trie.rs [hunk 21]
```diff
@@ -0,0 +1,1058 @@
...
+	/// If the trie does not contain a value for key, the returned proof contains all
+	/// nodes of the longest existing prefix of the key (at least the root node), ending
+	/// with the node that proves the absence of the key.
+	fn get_proof(&self, key: &[u8]) -> TrieResult<Proof>;
+
+	/// return value if key exists, None if key not exist, Error if proof is wrong
+	fn verify_proof(root_hash: Vec<u8>, key: &[u8], proof: Proof) -> TrieResult<Option<Vec<u8>>>;
+}
+
+impl Trie for MerklePatriciaTrie {
```
Comment (AurevoirXavier) [inline core/merkle-patricia-trie/src/trie.rs]: recommend add `#[inline]` for these tiny & light functions

File: core/sr-eth-primitives/src/lib.rs [hunk 22]
```diff
@@ -15,23 +15,9 @@ pub mod receipt;
 
 use codec::{Decode, Encode};
```
Comment (AurevoirXavier) [inline core/sr-eth-primitives/src/lib.rs]: update format

```rust
#![recursion_limit = "128"]
#![cfg_attr(not(feature = "std"), no_std)]

#[macro_use]
pub extern crate rlp_derive;

pub mod encoded;
pub mod error;
pub mod header;
pub mod pow;
pub mod receipt;
//pub mod transaction;

pub use ethbloom::{Bloom, Input as BloomInput};
pub use ethereum_types::H64;
pub use primitive_types::{H160, H256, U128, U256, U512};

pub type Bytes = Vec<u8>;
pub type Address = H160;
pub type BlockNumber = u64;

use codec::{Decode, Encode};
use rstd::vec::Vec;
```

File: core/merkle-patricia-trie/src/db.rs (hunk 5 above)
Comment (HackFisher) [reply core/merkle-patricia-trie/src/db.rs]: Using of ```#[inline]``` seems to be controversial here? Here is some discussion on this 
https://internals.rust-lang.org/t/when-should-i-use-inline/598/3

This crate is is strongly inspired by https://github.com/citahub/cita-trie , adding some proof related codes.

Comment (AurevoirXavier) [reply core/merkle-patricia-trie/src/db.rs]: https://stackoverflow.com/questions/37639276/when-should-inline-be-used-in-rust

Ok, ignore for the time being.

**Pull request merged**